File: 2025/bootstrap.swift
#!/usr/bin/env swift

import Foundation

enum TemplateProcessorError: Error, CustomStringConvertible {
    case incorrectArguments
    case sourceFolderNotFound(String)
    case targetLocationNotFound(String)
    case noValidReplacements(String)
    case copyFailed(Error)

    var description: String {
        switch self {
        case .incorrectArguments:
            return "Error: Incorrect number of arguments"
        case .sourceFolderNotFound(let path):
            return "Error: Source folder '\(path)' does not exist"
        case .targetLocationNotFound(let path):
            return "Error: Target location '\(path)' does not exist"
        case .noValidReplacements(let input):
            return "Error: No valid replacement pairs found in '\(input)'"
        case .copyFailed(let error):
            return "Error copying folder: \(error)"
        }
    }
}

struct ReplacementPair {
    let from: String
    let to: String
}

struct TemplateProcessor {
    private let fileManager = FileManager.default

    static func main() {
        let processor = TemplateProcessor()
        do {
            try processor.run()
        } catch {
            print(error)
            if error is TemplateProcessorError {
                processor.printUsage()
            }
            exit(1)
        }
    }

    func run() throws {
        let args = Array(CommandLine.arguments.dropFirst())
        let (sourceFolder, replacements, targetLocation) = try validateArgs(args)
        try validatePaths(sourceFolder: sourceFolder, targetLocation: targetLocation)
        let replacementPairs = try parseReplacements(replacements)
        try processTemplate(
            sourceFolder: sourceFolder,
            replacementPairs: replacementPairs,
            targetLocation: targetLocation
        )
    }

    private func validateArgs(_ args: [String]) throws -> (String, String, String) {
        guard args.count >= 3 else {
            throw TemplateProcessorError.incorrectArguments
        }
        return (args[0], args[1], args[2])
    }

    private func validatePaths(sourceFolder: String, targetLocation: String) throws {
        guard fileManager.fileExists(atPath: sourceFolder) else {
            throw TemplateProcessorError.sourceFolderNotFound(sourceFolder)
        }

        guard fileManager.fileExists(atPath: targetLocation) else {
            throw TemplateProcessorError.targetLocationNotFound(targetLocation)
        }
    }

    private func parseReplacements(_ replacementsStr: String) throws -> [ReplacementPair] {
        let pairs = replacementsStr.components(separatedBy: " ")
            .filter { $0.contains(",") }
            .compactMap { pair -> ReplacementPair? in
                let components = pair.components(separatedBy: ",")
                guard !components.isEmpty else { return nil }
                let from = components[0]
                let to = components.count > 1 ? components[1] : ""
                return ReplacementPair(from: from, to: to)
            }

        guard !pairs.isEmpty else {
            throw TemplateProcessorError.noValidReplacements(replacementsStr)
        }

        return pairs
    }

    private func processTemplate(
        sourceFolder: String,
        replacementPairs: [ReplacementPair],
        targetLocation: String
    ) throws {
        let newFolderName = replacementPairs.reduce(URL(fileURLWithPath: sourceFolder).lastPathComponent) { result, pair in
            result.replacingOccurrences(of: pair.from, with: pair.to)
        }

        let targetURL = URL(fileURLWithPath: targetLocation).appendingPathComponent(newFolderName)
        let targetPath = targetURL.path

        do {
            try fileManager.copyItem(atPath: sourceFolder, toPath: targetPath)
        } catch {
            throw TemplateProcessorError.copyFailed(error)
        }

        replacementPairs.forEach { pair in
            applyReplacementsToDirectories(targetPath, pair)
        }

        replacementPairs.forEach { pair in
            applyReplacementsToFilenames(targetPath, pair)
        }

        replacementPairs.forEach { pair in
            applyReplacementsToContents(targetPath, pair)
        }

        printSuccessMessage(sourceFolder: sourceFolder, targetPath: targetPath, replacementPairs: replacementPairs)
    }

    private func applyReplacementsToDirectories(_ targetPath: String, _ pair: ReplacementPair) {
        let targetURL = URL(fileURLWithPath: targetPath)

        guard let enumerator = fileManager.enumerator(
            at: targetURL,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else { return }

        let directories = enumerator.compactMap { element -> URL? in
            guard let url = element as? URL else { return nil }

            do {
                let resourceValues = try url.resourceValues(forKeys: [.isDirectoryKey])
                if resourceValues.isDirectory == true && url.lastPathComponent.contains(pair.from) {
                    return url
                }
            } catch {
                // Skip if we can't get resource values
            }

            return nil
        }

        // Sort directories by depth (deepest first) to avoid path issues
        let sortedDirectories = directories.sorted { url1, url2 in
            url1.pathComponents.count > url2.pathComponents.count
        }

        for directory in sortedDirectories {
            let newPath = directory.path.replacingOccurrences(of: pair.from, with: pair.to)
            guard directory.path != newPath else { continue }

            do {
                try fileManager.moveItem(atPath: directory.path, toPath: newPath)
            } catch {
                print("Error renaming directory: \(error)")
            }
        }
    }

    private func applyReplacementsToFilenames(_ targetPath: String, _ pair: ReplacementPair) {
        let targetURL = URL(fileURLWithPath: targetPath)

        guard let enumerator = fileManager.enumerator(
            at: targetURL,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else { return }

        let files = enumerator.compactMap { element -> URL? in
            guard let url = element as? URL else { return nil }

            do {
                let resourceValues = try url.resourceValues(forKeys: [.isDirectoryKey])
                if resourceValues.isDirectory != true && url.lastPathComponent.contains(pair.from) {
                    return url
                }
            } catch {
                // Skip if we can't get resource values
            }

            return nil
        }

        for file in files {
            let newPath = file.path.replacingOccurrences(of: pair.from, with: pair.to)
            guard file.path != newPath else { continue }

            do {
                try fileManager.moveItem(atPath: file.path, toPath: newPath)
            } catch {
                print("Error renaming file: \(error)")
            }
        }
    }

    private func applyReplacementsToContents(_ targetPath: String, _ pair: ReplacementPair) {
        let targetURL = URL(fileURLWithPath: targetPath)

        guard let enumerator = fileManager.enumerator(
            at: targetURL,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else { return }

        enumerator.forEach { element in
            guard let url = element as? URL else { return }

            do {
                let resourceValues = try url.resourceValues(forKeys: [.isDirectoryKey])
                guard resourceValues.isDirectory != true else { return }

                let content = try String(contentsOf: url, encoding: .utf8)
                let updatedContent = content.replacingOccurrences(of: pair.from, with: pair.to)
                try updatedContent.write(to: url, atomically: true, encoding: .utf8)
            } catch {
                // Skip binary files or files that can't be read as text
            }
        }
    }

    private func printUsage() {
        print("Usage: swift script.swift <source_folder> <replacements> <target_location>")
        print("Example: swift script.swift template_folder 'OLD_NAME,NEW_NAME FOO,BAR' .")
        print("Replacements format: 'FROM1,TO1 FROM2,TO2 FROM3,TO3'")
    }

    private func printSuccessMessage(
        sourceFolder: String,
        targetPath: String,
        replacementPairs: [ReplacementPair]
    ) {
        print("Successfully copied '\(sourceFolder)' to '\(targetPath)' with string replacements")
        replacementPairs.forEach { pair in
            print("Replaced '\(pair.from)' with '\(pair.to)' in folder names, file names, and file contents")
        }
    }
}

// Run the main function
TemplateProcessor.main()

File: 2025/bootstrap.exs
#!/usr/bin/env elixir

defmodule TemplateProcessor do
  @moduledoc """
  A script to copy a template folder and apply string replacements to folder names,
  file names, and file contents.
  """

  def main(args) do
    case validate_args(args) do
      {:ok, {source_folder, replacements, target_location}} ->
        case validate_paths(source_folder, target_location) do
          :ok ->
            case parse_replacements(replacements) do
              {:ok, replacement_pairs} ->
                process_template(source_folder, replacement_pairs, target_location)
              {:error, message} ->
                IO.puts(message)
                System.halt(1)
            end
          {:error, message} ->
            IO.puts(message)
            System.halt(1)
        end
      {:error, message} ->
        IO.puts(message)
        print_usage()
        System.halt(1)
    end
  end

  defp validate_args(args) do
    if length(args) < 3 do
      {:error, "Error: Incorrect number of arguments"}
    else
      [source_folder, replacements, target_location | _] = args
      {:ok, {source_folder, replacements, target_location}}
    end
  end

  defp validate_paths(source_folder, target_location) do
    cond do
      not File.exists?(source_folder) ->
        {:error, "Error: Source folder '#{source_folder}' does not exist"}
      not File.exists?(target_location) ->
        {:error, "Error: Target location '#{target_location}' does not exist"}
      true ->
        :ok
    end
  end

  defp parse_replacements(replacements_str) do
    pairs =
      replacements_str
      |> String.split(" ")
      |> Enum.filter(&String.contains?(&1, ","))

    if Enum.empty?(pairs) do
      {:error, "Error: No valid replacement pairs found in '#{replacements_str}'"}
    else
      replacement_pairs =
        Enum.map(pairs, fn pair ->
          [from, to] = String.split(pair, ",", parts: 2)
          {from, to}
        end)
      {:ok, replacement_pairs}
    end
  end

  defp process_template(source_folder, replacement_pairs, target_location) do
    # Create new folder name by applying all replacements to source folder name
    new_folder_name =
      Enum.reduce(replacement_pairs, Path.basename(source_folder), fn {from, to}, acc ->
        String.replace(acc, from, to)
      end)

    target_path = Path.join(target_location, new_folder_name)

    # Copy the entire folder structure to target location
    File.cp_r!(source_folder, target_path)

    # Apply all replacements to directory names (bottom-up to avoid path issues)
    Enum.each(replacement_pairs, fn {from_str, to_str} ->
      apply_replacements_to_directories(target_path, from_str, to_str)
    end)

    # Apply all replacements to file names
    Enum.each(replacement_pairs, fn {from_str, to_str} ->
      apply_replacements_to_filenames(target_path, from_str, to_str)
    end)

    # Apply all replacements to file contents
    Enum.each(replacement_pairs, fn {from_str, to_str} ->
      apply_replacements_to_contents(target_path, from_str, to_str)
    end)

    print_success_message(source_folder, target_path, replacement_pairs)
  end

  defp apply_replacements_to_directories(target_path, from_str, to_str) do
    Path.wildcard(Path.join(target_path, "**/*#{from_str}*"))
    |> Enum.filter(&File.dir?/1)
    |> Enum.each(fn dir ->
      new_dir = String.replace(dir, from_str, to_str)
      if dir != new_dir do
        File.rename!(dir, new_dir)
      end
    end)
  end

  defp apply_replacements_to_filenames(target_path, from_str, to_str) do
    Path.wildcard(Path.join(target_path, "**/*#{from_str}*"))
    |> Enum.filter(&File.regular?/1)
    |> Enum.each(fn file ->
      new_file = String.replace(file, from_str, to_str)
      if file != new_file do
        File.rename!(file, new_file)
      end
    end)
  end

  defp apply_replacements_to_contents(target_path, from_str, to_str) do
    Path.wildcard(Path.join(target_path, "**/*"))
    |> Enum.filter(&File.regular?/1)
    |> Enum.each(fn file ->
      try do
        content = File.read!(file)
        updated_content = String.replace(content, from_str, to_str)
        File.write!(file, updated_content)
      rescue
        # Skip binary files or files that can't be read as text
        _ -> :ok
      end
    end)
  end

  defp print_usage do
    IO.puts("Usage: elixir script.exs <source_folder> <replacements> <target_location>")
    IO.puts("Example: elixir script.exs template_folder 'OLD_NAME,NEW_NAME FOO,BAR' .")
    IO.puts("Replacements format: 'FROM1,TO1 FROM2,TO2 FROM3,TO3'")
  end

  defp print_success_message(source_folder, target_path, replacement_pairs) do
    IO.puts("Successfully copied '#{source_folder}' to '#{target_path}' with string replacements")
    Enum.each(replacement_pairs, fn {from, to} ->
      IO.puts("Replaced '#{from}' with '#{to}' in folder names, file names, and file contents")
    end)
  end
end

# Run the main function with command line arguments
System.argv() |> TemplateProcessor.main()
